files/scripts/arp_update

```diff
@@ -89,32 +89,57 @@ while /bin/true; do
+            eval `eval "$ip_neigh_flush_cmd"`
+        fi
+
+        # generates the following command for each FAILED or INCOMPLETE IPv6 neighbor
         # timeout 0.2 ping <neighbor IPv6> -n -q -i 0 -c 1 -W 1 -I <VLAN name> >/dev/null
-        ping6_template="sed -e 's/^/timeout 0.2 ping /' -e 's/,/ -n -q -i 0 -c 1 -W 1 -I /' -e 's/$/ >\/dev\/null;/'"
-        failed_ip6_neigh_cmd="ip -6 neigh show | grep -v fe80 | grep $vlan | grep -E 'FAILED|INCOMPLETE' | cut -d ' ' -f 1,3 --output-delimiter=',' | $ping6_template"
-        eval `eval $failed_ip6_neigh_cmd`
+        if [[ ! -z "$unresolved_kernel_neighbors" ]]; then
+            ping6_template="sed -e 's/^/timeout 0.2 ping /' -e 's/,/ -n -q -i 0 -c 1 -W 1 -I /' -e 's/$/ >\/dev\/null;/'"
+            failed_ip6_neigh_cmd="echo \"$unresolved_kernel_neighbors\" | cut -d ' ' -f 1,3 --output-delimiter=',' | $ping6_template"
+            eval `eval "$failed_ip6_neigh_cmd"`
+        fi
+
+        # wait for any transient INCOMPLETE neighbors to transition to FAILED
+        # once these transition to FAILED, we can guarantee that the kernel has generated a netlink message for the neighbor
+        # and proceed setting it to permanent INCOMPLETE
+        # this is necessary for:
+        # 1. unsyncronized neighbors that were flushed, then pinged
+        # 2. existing FAILED neighbors which will be set to transient INCOMPLETE by the above ping
+        wait_neighbors="$unsync_neighbors"$'\n'"$failed_kernel_neighbors"
+        if [[ ! -z "$wait_neighbors" ]]; then
+            trans_incomplete_check=$(echo "$wait_neighbors" | sed  -e 's/^/ip neigh show /' -e 's/$/ | grep -q 'INCOMPLETE' \&\&/')
+            trans_incomplete_check=$(echo $trans_incomplete_check | sed -e 's/\&\&$//')
+            timeout 10 bash -c "while eval \"$trans_incomplete_check\"; do echo 'waiting'; sleep 1; done"
+        fi
+
+        # manually set any remaining FAILED entries to permanently INCOMPLETE
+        # once these entries are INCOMPLETE, any subsequent neighbor advertisement messages are able to resolve the entry
+        # ignore INCOMPLETE neighbors since if they are transiently incomplete (i.e. new kernel neighbors that we are attempting to resolve for the first time),
+        # setting them to permanently incomplete here means the kernel will never generate a netlink message for that neighbor
+        # generates the following command for each FAILED IPv6 neighbor
+        # ip neigh replace <neighbor IPv6> dev <VLAN name> nud incomplete
+        failed_kernel_neighbors=$(ip -6 neigh show | grep -v fe80 | grep $vlan | grep -E 'FAILED')
+        if [[ ! -z "$failed_kernel_neighbors" ]]; then
+            neigh_replace_template="sed -e 's/^/ip neigh replace /' -e 's/,/ dev /' -e 's/$/ nud incomplete;/'" 
+            ip_neigh_replace_cmd="echo \"$failed_kernel_neighbors\" | cut -d ' ' -f 1,3 --output-delimiter=',' | $neigh_replace_template"
+            eval `eval "$ip_neigh_replace_cmd"`
```

Comment:
can you please test this with traffic which can resolve a neighbor during a ping and another where neighbor cannot be resolved?

Reply:
Tested with one resolvable IP, one unresolvable IP, and one unresolvable IP where I manually deleted the APPL_DB entry. The script behaved as expected - the deleted IP was flushed, all three IPs were pinged, and the final state of the IPs was REACHABLE (for the resolvable IP) and INCOMPLETE for the other two.